### PR TITLE
DBZ-2726 Clean MySQL TIMESTAMP default value

### PR DESCRIPTION
**Pull request by keweishang**
https://issues.redhat.com/browse/DBZ-2726

Although MySQL can do some basic validation of the default value of the `TIMESTAMP` column, all the following default values can be applied successfully to MySQL, because MySQL can automatically fix them when applying the DDL. However, the raw DDL is sent to Debezium and many of the values will make the MySQL Connector task fail when updating the internal schema:

```sql
CREATE TABLE my_table (
    ts_col01 TIMESTAMP DEFAULT '2020-01-02',
    ts_col02 TIMESTAMP DEFAULT '2020-01-02 ',
    ts_col03 TIMESTAMP DEFAULT '2020-01-02:',
    ts_col04 TIMESTAMP DEFAULT '2020-01-02--',
    ts_col05 TIMESTAMP DEFAULT '2020-01-02 03',
    ts_col06 TIMESTAMP DEFAULT '2020-01-02 003',
    ts_col07 TIMESTAMP DEFAULT '2020-01-02 03:',
    ts_col08 TIMESTAMP DEFAULT '2020-01-02 03:04',
    ts_col09 TIMESTAMP DEFAULT '2020-01-02 03:004',
    ts_col10 TIMESTAMP DEFAULT '2020-01-02 03:04:05',
    ts_col11 TIMESTAMP DEFAULT '2020-01-02 03:04:05.6',
    ts_col12 TIMESTAMP DEFAULT '2020-01-02:03:04:05',
    ts_col13 TIMESTAMP DEFAULT '2020-01-02-03:04:05',
    ts_col14 TIMESTAMP DEFAULT '2020-01-02--03:04:05',
    ts_col15 TIMESTAMP DEFAULT '2020-01-02--03:004:0005',
    ts_col16 TIMESTAMP DEFAULT '02020-0001-00002--03:004:0005',
    ts_col17 TIMESTAMP DEFAULT '1970-01-01:00:00:001'
);
```

Proposed solution:
Clean the TIMESTAMP default value passed from the DDL.

### CODE REVIEW

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlDefaultValueConverter.java [hunk 1]
```diff
@@ -259,6 +260,125 @@ private DateTimeFormatter timestampFormat(int length) {
...
+                                second = Integer.parseInt(s.substring(secondColon + 1, period));
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        StringBuilder cleanedTimestamp = new StringBuilder();
+        cleanedTimestamp = cleanedTimestamp
```
Comment (jpechane) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlDefaultValueConverter.java:362]: Could this part be consturcted with a single `String.format` for all compoents?

File: debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/MySqlAntlrDdlParserTest.java [hunk 2]
```diff
@@ -2542,6 +2544,87 @@ public void shouldNotNullPositionBeforeOrAfterDefaultValue() {
...
+                "ts_col02 TIMESTAMP DEFAULT '2020-01-02 '," +
+                "ts_col03 TIMESTAMP DEFAULT '2020-01-02:'," +
+                "ts_col04 TIMESTAMP DEFAULT '2020-01-02--'," +
+                "ts_col05 TIMESTAMP DEFAULT '2020-01-02 03'," +
+                "ts_col06 TIMESTAMP DEFAULT '2020-01-02 003'," +
+                "ts_col07 TIMESTAMP DEFAULT '2020-01-02 03:'," +
+                "ts_col08 TIMESTAMP DEFAULT '2020-01-02 03:04'," +
+                "ts_col09 TIMESTAMP DEFAULT '2020-01-02 03:004'," +
+                "ts_col10 TIMESTAMP DEFAULT '2020-01-02 03:04:05'," +
+                "ts_col11 TIMESTAMP DEFAULT '2020-01-02 03:04:05.6'," +
```
Comment (jpechane) [inline debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/MySqlAntlrDdlParserTest.java:2562]: Is also acceptable case when there is the period but no following numbers?

**Comment by jpechane** [comment]
@keweishang Hi, this is great work! Unfoirtunately it seems that MySQL is even more perverse in this regard :-(
I tried few more cases
```
2001-12-23 12::30
2001-12-23 12::30.
2001-12-23 12::30.12
2001-12-23 12.30
2001#12#23 12.30
2001#12-23^12.30
```
And all of them are considered valid :-(

It seems to me that either there is no seprator or arbitrary char could be used as separator.
Could you please take a look and exted the logic?

**Comment by keweishang** [comment]
@jpechane Thanks for reviewing! Indeed, MySQL is even more perverse in this regard. I'll check these cases and extend the logic.

File: debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/MySqlAntlrDdlParserTest.java (hunk 2 above)
Comment (keweishang) [reply debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/MySqlAntlrDdlParserTest.java:2562]: It is an acceptable case both in MySQL and in Debezium when no numbers follow the period. I just added a test case for it.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlDefaultValueConverter.java (hunk 1 above)
Comment (keweishang) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlDefaultValueConverter.java:362]: Absolutely! Addressed.

**Comment by keweishang** [comment]
@jpechane I've extended the logic to accept the following cases that MySQL tolerates:
```
2001-12-23 12::30
2001-12-23 12::30.
2001-12-23 12::30.12
2001-12-23 12.30
2001#12#23 12.30
2001##12--23^12.30
```

**Comment by jpechane** [comment]
@keweishang Applied, thanks!

**Comment by keweishang** [comment]
Thanks @jpechane 👍

**Pull request merged**